README.md


# Tanaka <img src="man/figures/logo.png" align="right" width="140"/>

[![Cran version](https://www.r-pkg.org/badges/version-ago/tanaka)](https://CRAN.R-project.org/package=tanaka)
[![R-CMD-check](https://github.com/riatelab/tanaka/workflows/R-CMD-check/badge.svg)](https://github.com/riatelab/tanaka/actions)
[![Coverage status](https://codecov.io/gh/riatelab/tanaka/branch/master/graph/badge.svg)](https://app.codecov.io/github/riatelab/tanaka?branch=master)
![N. Download](https://cranlogs.r-pkg.org/badges/grand-total/tanaka?color=brightgreen)  
[![Project Status: Active – The project has reached a stable, usable
state and is being actively
developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)



Also called "relief contours method", "illuminated contour method" or "shaded 
contour lines method", the Tanaka method<sup>[1](#fn1)</sup> enhances the representation of topography 
on a map by using shaded contour lines. The result is a 3D-like map.

This package is a simplified implementation of the Tanaka method, north-west white contours represent 
illuminated topography and south-east black contours represent shaded topography. 
Even if the results are quite satisfactory, a more refined method could be used 
based on the Kennelly and Kimerling's paper<sup>[2](#fn2)</sup>. 


`tanaka` is a small package with two functions:

- `tanaka()` uses a `terra` object and displays the map directly;
- `tanaka_contour()` builds the isopleth polygon layer. 


The contour lines creation relies on [`mapiso`](https://github.com/riatelab/mapiso), 
spatial manipulation and display rely on [`sf`](https://github.com/r-spatial/sf). 


## Installation
* From CRAN
```r
install.packages("tanaka")
```

* Development version on GitHub
```r
require(remotes)
install_github("riatelab/tanaka")
```

## Demo

* This example is based on the dataset shipped within the package. 
```r
library(tanaka)
library(terra)
ras <- rast(system.file("tif/elev.tif", package = "tanaka"))
tanaka(ras, breaks = seq(80,400,20), 
       legend.pos = "topright", legend.title = "Elevation\n(meters)")
```
![](https://raw.githubusercontent.com/riatelab/tanaka/master/img/ex1.png)  

* This example is based on an  elevation raster downloaded via 
[`elevatr`](https://github.com/jhollist/elevatr). 
```r
library(tanaka)
library(elevatr)
library(terra)
# use elevatr to get elevation data
ras <- get_elev_raster(locations = data.frame(x = c(6.7, 7), y = c(45.8,46)),
                       z = 10, prj = "EPSG:4326", clip = "locations")
ras <- rast(ras)
# custom color palette
cols <- c("#F7E1C6", "#EED4C1", "#E5C9BE", "#DCBEBA", "#D3B3B6", "#CAA8B3", 
          "#C19CAF", "#B790AB", "#AC81A7", "#A073A1", "#95639D", "#885497", 
          "#7C4692", "#6B3D86", "#573775", "#433266", "#2F2C56", "#1B2847")
# display the map
tanaka(ras, breaks = seq(500,4800,250), col = cols)
```
![](https://raw.githubusercontent.com/riatelab/tanaka/master/img/ex2.png)  

* The last example illustrates the use of tanaka with non-topographical data. 
This map is based on the [Global Human Settlement Population Grid](https://ghsl.jrc.ec.europa.eu/ghs_pop.php) (1km). 

```r
library(terra)
library(sf)
library(tanaka)
library(mapsf)
# Download
tempzip <- tempfile()
tempfolder <- tempdir()
data_url <- paste0("http://cidportal.jrc.ec.europa.eu/ftp/jrc-opendata/GHSL/", 
                   "GHS_POP_GPW4_GLOBE_R2015A/GHS_POP_GPW42015_GLOBE_R2015A_54009_1k/", 
                   "V1-0/GHS_POP_GPW42015_GLOBE_R2015A_54009_1k_v1_0.zip")
download.file(data_url, tempzip)
unzip(tempzip, exdir = tempfolder)
# Import
pop2015 <- rast(paste0(tempfolder, 
                       "/GHS_POP_GPW42015_GLOBE_R2015A_54009_1k_v1_0/",
                       "GHS_POP_GPW42015_GLOBE_R2015A_54009_1k_v1_0.tif"))
# Mask raster
center <- st_as_sf(data.frame(x=425483.8, y=5608290), 
                   coords=(c("x","y")), 
                   crs = st_crs(pop2015))
center <- st_buffer(center, dist = 800000)
ras <- crop(pop2015, st_bbox(center)[c(1,3,2,4)])
# Smooth values
mat <- focalMat(x = ras, d = c(10000), type = "Gauss")
rassmooth <- focal(x = ras, w = mat, fun = sum, na.rm = TRUE)
# Map
bks <- c(0,25,50,100,250,500,750,1000,1750,2500,5000, 7500,10000)
mf_export(x = center, filename = "circle.png", width = 800, res = 100)
tanaka(x = rassmooth,
       breaks = bks,
       mask = center,
       col = hcl.colors(n = 12, palette = "Inferno"),
       shift = 2500,
       add = TRUE,
       legend.pos = "topleft",
       legend.title = "Inhabitants\nper km2")
mf_map(center, add = TRUE, border = "white", col = NA, lwd = 6)
mf_title(txt = "Smoothed Population Density")
mf_credits(paste0("T. Giraud, 2022\n",
                  "Data : European Commission, Joint Research Centre (JRC); ", 
                  "Columbia University, CIESIN (2015): GHS population grid, ", 
                  "derived from GPW4."))
text(-250000 ,6420000, "Gaussian filter, sigma = 10km", 
     adj = 0, font = 3, cex = .8, xpd = TRUE )
dev.off()
```


![](https://raw.githubusercontent.com/riatelab/tanaka/master/img/circle.png)

## Alternative Package
The [`metR` package](https://CRAN.R-project.org/package=metR) allows to draw [Tanaka contours with ggplot2](https://eliocamp.github.io/metR/reference/geom_contour_tanaka.html).


-------------------------------------------

<a name="fn1">1</a>: Tanaka, K. (1950). The relief contour method of representing topography on maps. *Geographical Review, 40*(3), 444-456.  
<a name="fn2">2</a>: Kennelly, P., & Kimerling, A. J. (2001). Modifications of Tanaka's illuminated contour method. *Cartography and Geographic Information Science, 28*(2), 111-123.




NEWS.md
# tanaka 0.4.0

## Minor changes
- use maplegend for legends


# tanaka 0.3.0

## Minor changes
- replace isoband by mapiso
- use inherits() instead of is() (remove methods from Imports)

# tanaka 0.2.0

## Major changes

- We use terra instead of raster for all things raster
- We use tinytest instead of testthat


# tanaka 0.1.3

## Minor changes

- Changes to accommodate sf >= 0.9 and st_make_valid no longer in lwgeom



# tanaka 0.1.2

## Minor changes

- Changes to accomodate sf >= 0.9 and new representations in PROJ. 




# tanaka 0.1.1

## Minor changes

- addition of `add` argument in tanaka() to add a tanaka layer to and existing
plot.   
- addtition of reference to Tanaka paper. 
- better description of the package (?tanaka)
- add param to correct north arrow
- cleaner raster to matrix transformation in tanaka_contour()
- add reference to GH repo for URL and issues
- use lintr and goodpractice





data-raw/test.R
# library(sf)
# com <- st_read("/home/tim/Téléchargements/ADMIN-EXPRESS-COG_1-1__SHP__FRA_2018-04-03/ADMIN-EXPRESS-COG/1_DONNEES_LIVRAISON_2018-03-28/ADE-COG_1-1_SHP_LAMB93_FR/COMMUNE_CARTO.shp")
# com46 <- com[com$INSEE_COM %in% c(46185,46106, 46330,46265, 46058, 46028, 46084,46208),]
# library(elevatr)
# ras0 <- get_elev_raster(com46, z=10)
# library(raster)
# ras0 <- crop(ras0, com46)
# writeRaster(x = ras0, filename = "inst/grd/elev.grd")
# st_write(com46, "inst/gpkg/com.gpkg", delete_dsn = T)


library(tanaka)
library(raster)
library(cartography)
library(sf)
library(elevatr)

ras <- raster(system.file("grd/elev.grd", package = "tanaka"))
ras <- crop(ras, extent(586101.7,603178.4, 6421500, 6428100))
iso <- tanaka_contour(ras, breaks = seq(80,400,20))
getFigDim(x = iso, width = 700, mar = c(0,0,0,0), res = 100)
png("img/banner.png", width = 700, height = 270, res =100, bg = NA)
par(mar = c(0,0,0,0))
tanaka(iso,legend.pos = "n")
dev.off()

ras <- raster(system.file("grd/elev.grd", package = "tanaka"))
png("img/ex1.png", width = 700, height = 533, res =100, bg = NA)
par(mar = c(0,0,0,0))
tanaka(ras, breaks = seq(80,400,20),
       legend.pos = "topright", legend.title = "Elevation\n(meters)")
dev.off()

# use elevatr to get elevation data
ras <- get_elev_raster(locations = data.frame(x = c(6.7, 7), y = c(45.8,46)),
                       z = 10, prj = "+init=epsg:4326", clip = "locations")
# custom color palette
cols <- c("#F7E1C6", "#EED4C1", "#E5C9BE", "#DCBEBA", "#D3B3B6", "#CAA8B3",
          "#C19CAF", "#B790AB", "#AC81A7", "#A073A1", "#95639D", "#885497",
          "#7C4692", "#6B3D86", "#573775", "#433266", "#2F2C56", "#1B2847")
png("img/ex2.png", width = 700, height = 533, res =100, bg = NA)
# display the map
par(mar=c(0,0,0,0))
tanaka(ras, breaks = seq(500,4800,250), col = cols)
dev.off()



tools::toTitleCase("Design Shaded Contour Lines (or Tanaka) Maps")


tests/testthat.R
library(testthat)
library(tanaka)
library(raster)
library(sf)



test_check("tanaka")


tests/tinytest.R
if (requireNamespace("tinytest", quietly = TRUE)) {
  tinytest::test_package("tanaka")
}


R/legend.R
## this function is a light version of legendChoro from cartography...
#' @name legendtanaka
#' @title legendtanaka
#' @importFrom graphics par rect strheight strwidth xinch text
#' @noRd
legendtanaka <- function(pos = "topleft",
                         title.txt = "Title of the legend",
                         title.cex = 0.8,
                         values.cex = 0.6,
                         breaks,
                         col,
                         cex = 1,
                         values.rnd = 2,
                         border = "black") {
  # exit for none
  positions <- c(
    "bottomleft",
    "topleft",
    "topright",
    "bottomright",
    "left",
    "right",
    "top",
    "bottom",
    "center",
    "bottomleftextra"
  )
  if (length(pos) == 1) {
    if (!pos %in% positions) {
      return(invisible())
    }
  }

  # figdim in geo coordinates
  x1 <- par()$usr[1]
  x2 <- par()$usr[2]
  y1 <- par()$usr[3]
  y2 <- par()$usr[4]

  # offsets
  delta1 <- xinch(0.15) * cex
  delta2 <- delta1 / 2

  # variables internes
  width <- (x2 - x1) / (30 / cex)
  height <- width / 1.5

  # extent
  if (!is.character(breaks)) {
    breaks <- as.numeric(round(breaks, values.rnd))
  }

  nodata.txt <- NULL
  longval <- max(strwidth(c(breaks, nodata.txt), cex = values.cex))
  legend_xsize <- max(
    width + longval,
    strwidth(title.txt, cex = title.cex) - delta2
  ) - delta2
  legend_ysize <-
    (length(breaks) - 1) * height + strheight(title.txt,
      cex = title.cex
    )

  # Get legend position
  legcoord <- legpos(
    pos = pos,
    x1 = x1,
    x2 = x2,
    y1 = y1,
    y2 = y2,
    delta1 = delta1,
    delta2 = delta2,
    legend_xsize = legend_xsize,
    legend_ysize = legend_ysize
  )
  xref <- legcoord$xref
  yref <- legcoord$yref

  # box display
  for (i in 0:(length(breaks) - 2)) {
    rect(
      xref,
      yref + i * height,
      xref + width,
      yref + height + i * height,
      col = col[i + 1],
      border = border,
      lwd = 0.4
    )
  }


  # text display
  for (i in 1:(length(breaks))) {
    text(
      x = xref + width + delta2,
      y = yref + (i - 1) * height,
      labels = breaks[i],
      adj = c(0, 0.5),
      cex = values.cex
    )
  }

  # title
  text(
    x = xref,
    y = yref + (length(breaks) - 1) * height + delta1,
    labels = title.txt,
    adj = c(0, 0),
    cex = title.cex
  )
}



legpos <- function(pos,
                   x1,
                   x2,
                   y1,
                   y2,
                   delta1,
                   delta2,
                   legend_xsize,
                   legend_ysize) {
  # Position
  if (length(pos) == 2) {
    return(list(xref = pos[1], yref = pos[2]))
  }
  if (pos == "bottomleft") {
    xref <- x1 + delta1
    yref <- y1 + delta1
  }
  if (pos == "bottomleftextra") {
    xref <- x1 + delta1
    yref <-
      y1 + delta1 + graphics::strheight(s = "hp\nhp", cex = 0.6, font = 3)
  }
  if (pos == "topleft") {
    xref <- x1 + delta1
    yref <- y2 - 2 * delta1 - legend_ysize
  }
  if (pos == "topright") {
    xref <- x2 - 2 * delta1 - legend_xsize
    yref <- y2 - 2 * delta1 - legend_ysize
  }
  if (pos == "bottomright") {
    xref <- x2 - 2 * delta1 - legend_xsize
    yref <- y1 + delta1
  }
  if (pos == "left") {
    xref <- x1 + delta1
    yref <- (y1 + y2) / 2 - legend_ysize / 2 - delta2
  }
  if (pos == "right") {
    xref <- x2 - 2 * delta1 - legend_xsize
    yref <- (y1 + y2) / 2 - legend_ysize / 2 - delta2
  }
  if (pos == "top") {
    xref <- (x1 + x2) / 2 - legend_xsize / 2
    yref <- y2 - 2 * delta1 - legend_ysize
  }
  if (pos == "bottom") {
    xref <- (x1 + x2) / 2 - legend_xsize / 2
    yref <- y1 + delta1
  }
  if (pos == "center") {
    xref <- (x1 + x2) / 2 - legend_xsize / 2
    yref <- (y1 + y2) / 2 - legend_ysize / 2 - delta2
  }
  return(list(xref = xref, yref = yref))
}


R/contour.R
#' @title Contour Layer
#' @name tanaka_contour
#' @description Create a contour layer.
#' @param x a raster or an sf contour layer.
#' @param nclass number of class.
#' @param breaks list of breaks.
#' @param mask a mask layer, sf object.
#' @export
#' @import sf
#' @import isoband
#' @import raster
#' @import lwgeom
#' @examples
#' library(tanaka)
#' library(raster)
#' library(sf)
#' ras <- raster(system.file("grd/elev.grd", package = "tanaka"))
#' iso <- tanaka_contour(x = ras)
#' plot(st_geometry(iso), col = c("#FBDEE1", "#F0BFC3", "#E7A1A6",
#'                                "#DD8287", "#D05A60", "#C03239",
#'                                "#721B20", "#1D0809"))
tanaka_contour <- function(x, nclass = 8, breaks, mask){
  ext <- x@extent
  nc <- x@ncols
  nr <- x@nrows
  val <- values(x)
  lon <- seq(ext[1], ext[2], length.out = nc)
  lat <- seq(ext[3], ext[4], length.out = nr)
  m <- matrix(data = val, nrow = nc, dimnames = list(lon,lat))
  vmin <- min(m, na.rm = TRUE)
  vmax <- max(m, na.rm = TRUE)
  if(missing(breaks)){
    breaks <- seq(from = vmin, to = vmax, length.out = (nclass+1))
  }else{
    breaks <- sort(unique(c(vmin, breaks[breaks > vmin & breaks < vmax], vmax)))
  }
  lev_low = breaks[1:(length(breaks)-1)]
  lev_high = breaks[2:length(breaks)]
  raw <- isobands(x = as.numeric(rownames(m)),
                  y = rev(as.numeric(colnames(m))),
                  z = t(m),
                  levels_low = lev_low,
                  levels_high = c(lev_high[-length(lev_high)], vmax + 1e-10))
  bands <- iso_to_sfg(raw)
  res <- st_sf(id = 1:length(bands),
               min = lev_low,
               max = lev_high,
               geometry = st_sfc(bands),
               crs = st_crs(x))
  # clean geoms
  st_geometry(res) <- st_make_valid(st_geometry(res))
  if(methods::is(st_geometry(res),"sfc_GEOMETRY")){
    st_geometry(res) <-   st_collection_extract(st_geometry(res), "POLYGON")
  }

  if(!missing(mask)){
    if(methods::is(mask, "Spatial")){mask <- st_as_sf(mask)}
    options(warn=-1)
    res <- st_cast(st_intersection(x = res, y = st_union(mask)))
    options(warn=0)
  }
  return(res)
}



data-raw/ex.R
library(sf)
library(spatstat)
library(sp)
library(maptools)
library(raster)
library(cartography)
library(SpatialPosition)

## import dataset
feat <- sf::st_read("https://gist.githubusercontent.com/rCarto/747164575e3f216a123c3092d0ce9162/raw/f12390464f255b5f9760c577ab6bf5456cf61a40/iris75.geojson")
# Use french projection
feat <- sf::st_transform(feat, 2154)

## Compute density raster
# from sf to sp
feat <- as(feat,'Spatial')
plot(feat)
# get coodinates
coords <- coordinates(feat)
# from sp to spatstat
pts <- ppp(x = coords[,1], y = coords[,2], window = as.owin(feat, 10),
           marks = feat$P14_POP)
# compute density
ds <- density.ppp(x = pts, sigma = 300, eps = 100, weights = pts$marks)
# spatstat to raster, in inhab per sq. km
ras <- raster(ds, crs = proj4string(feat)) * 1000 * 1000
ras[is.na(ras)] <- 0
feat <- st_as_sf(feat)
saveRDS(ras, "paris.rds")
saveRDS(feat, "mask_paris.rds")


mtq <- st_read(system.file("gpkg/mtq.gpkg", package="cartography"))
# use WGS84 proj
mtq_latlon <- st_transform(mtq, 4326)
# import raster
ras <- raster("srtm_24_10.tif")
# crop on martinique area
mtq_ras <- crop(ras, st_bbox(mtq_latlon)[c(1,3,2,4)])
# aggregate the raster
mtq_ras <- aggregate(mtq_ras, fact=4,fun=mean)
mtq_ras <- projectRaster(mtq_ras, crs = "+proj=utm +zone=20 +datum=WGS84 +units=m +no_defs")

saveRDS(mtq_ras, "mtq.rds")
saveRDS(mtq, "mask_mtq.rds")








data-raw/xx.R
library(raster)

rasto <- raster(x = "/home/tim/Téléchargements/GHS_POP_GPW42015_GLOBE_R2015A_54009_1k_v1_0/GHS_POP_GPW42015_GLOBE_R2015A_54009_1k_v1_0.tif")


library(cartography)
data(nuts2006)
m1 <- st_as_sf(nuts0.spdf)
m1 <- st_as_sf(nuts0.spdf)

m1 <- st_transform(x = m1, st_crs(rasto))
mx <- m1
m1 <- (m1[m1$id%in%c("FR", 'BE', 'DE',"LU", "CH"),])
m1$id
library(sf)
ras <- crop(x = rasto, y = m1)
library(rnaturalearth)


ctr <- (ne_countries(scale = 50,returnclass = "sf" )  )
mm <- st_transform(ctr, st_crs(m1))

plot(ras)
ras2 <- aggregate(ras, fact=7,fun=sum)
ras2 <- ras2/49

mat <- raster::focalWeight(x = ras, d = c(10000, 25000),type = "Gauss")
xx <- raster::focal(x = ras, w = mat, fun = sum,
                   na.rm = FALSE, pad = TRUE, padValue = 0)

plot(xx,
     col = carto.pal("wine.pal", 15))

getBreaks(v = c(1:9000), method = "geom")

bks <- getBreaks(v = values(xx)[values(xx)>1], nclass = 10, method = "geom")
bks <- c(0,50,100,250,500,1000,1500,2000,4000,8000,9000)
ras
par(mar = c(0,0,0,0))
bks <- c(0,50,100,250,500,1000,1500,2000,4000,8000,10000,20000, 100000000)
tanaka(xx, breaks = bks, shift = 1000, mask = mm, dark = "grey20",
       col = carto.pal("pink.pal", length(bks)-1))
plot(mm$geometry, add=T)

display.carto.all()


plot(m1$geometry, add=T)

(10000*10000) / (1000*1000)

library(linemap)
library(sf)
data("occitanie")
popOcc1 <- smoothgrid(x = popOcc, var = "pop", d = c(1000,5000), type = "Gauss")
opar <- par(mar=c(0,0,0,0), bg = "ivory2")
plot(st_geometry(occitanie), col="ivory1", border = NA)
linemap(x = popOcc1, var = "pop", k = 5, threshold = 25,
        col = "ivory1", border = "ivory4", lwd = 0.6, add = TRUE)
par(opar)


R/tanaka_contour.R
#' @title Create a Contour Layer
#' @name tanaka_contour
#' @description Create a contour layer.
#' @param x a raster object.
#' @param nclass a number of class.
#' @param breaks a vector of break values.
#' @param mask a mask layer, a POLYGON or MULTIPOLYGON sf object.
#' @return A MULTIPOLYGON sf object is return. The data.frame contains 3 fields:
#' id, min (minimum value of the raster in the MULTIPOLYGON) and max (maximum
#' value of the raster in the MULTIPOLYGON).
#' @export
#' @importFrom sf st_sf st_sfc st_geometry st_collection_extract st_crs st_agr<-
#' st_cast st_intersection st_union st_as_sf st_geometry<-
#' st_make_valid
#' @importFrom mapiso mapiso
#' @examples
#' library(tanaka)
#' library(terra)
#' library(sf)
#' ras <- rast(system.file("tif/elev.tif", package = "tanaka"))
#' iso <- tanaka_contour(x = ras)
#' plot(st_geometry(iso), col = c(
#'   "#FBDEE1", "#F0BFC3", "#E7A1A6",
#'   "#DD8287", "#D05A60", "#C03239",
#'   "#721B20", "#1D0809"
#' ))
tanaka_contour <- function(x, nclass = 8, breaks, mask) {
  iso <- mapiso(x = x, nbreaks = nclass, breaks = breaks, mask = mask)
  names(iso)[1:3] <- c("id", "min", "max")
  return(iso)
}


R/tanaka.R
#' @title Plot a Tanaka Map
#' @name tanaka
#' @description This function plots a tanaka map.
#' @param x a raster or an sf contour layer (e.g. the result of
#' \code{tanaka_contour()}, "min" and "max" columns are needed).
#' @param nclass a number of class.
#' @param breaks a vector of break values.
#' @param mask a mask layer, a POLYGON or MULTIPOLYGON sf object.
#' @param col a color palette (a vector of colors).
#' @param light light shadow (NW color).
#' @param dark dark shadow (SE color).
#' @param shift size of the shadow (in map units).
#' @param legend.pos position of the legend, one of 'topleft', 'top',
#' 'topright', 'right', 'bottomright', 'bottom', 'bottomleft',
#' 'left' or a vector of two coordinates in map units (c(x, y)).
#' If leg.position = NA then the legend is not plotted.
#' If leg.position = 'interactive' click on the map to choose the
#' legend position.
#' @param legend.title title of the legend.
#' @param add whether to add the layer to an existing plot (TRUE) or
#' not (FALSE).
#' @export
#' @importFrom sf st_geometry st_bbox
#' @importFrom grDevices colorRampPalette
#' @importFrom graphics plot
#' @importFrom terra crs
#' @importFrom maplegend leg
#' @references Tanaka, K. (1950). The relief contour method of representing
#' topography on maps. \emph{Geographical Review, 40}(3), 444-456.
#' @return A Tanaka contour map is plotted.
#' @examples
#' library(tanaka)
#' library(terra)
#' library(sf)
#' com <- st_read(system.file("gpkg/com.gpkg", package = "tanaka"),
#'   quiet = TRUE
#' )
#' ras <- rast(system.file("tif/elev.tif", package = "tanaka"))
#' tanaka(ras)
#' tanaka(ras, mask = com)
#' tanaka(ras,
#'   breaks = seq(80, 400, 20),
#'   legend.pos = "topright",
#'   legend.title = "Elevation\n(meters)"
#' )
#' tanaka(ras,
#'   nclass = 15,
#'   col = hcl.colors(15, "YlOrRd"),
#'   legend.pos = "topright",
#'   legend.title = "Elevation\n(meters)"
#' )
tanaka <- function(x,
                   nclass = 8,
                   breaks,
                   col,
                   mask,
                   light = "#ffffff70",
                   dark = "#00000090",
                   shift,
                   legend.pos = "left",
                   legend.title = "Elevation",
                   add = FALSE) {
  if (inherits(x = x, what = "SpatRaster")) {
    x <-
      tanaka_contour(
        x = x,
        nclass = nclass,
        breaks = breaks,
        mask = mask
      )
  }
  nclass <- nrow(x)
  if (missing(col)) {
    palet <-
      c(
        "#FBDEE1",
        "#F7D2D6",
        "#F3C7CB",
        "#F0BCC0",
        "#ECB1B5",
        "#E9A6AB",
        "#E59BA0",
        "#E29095",
        "#DE858A",
        "#D9767B",
        "#D5676D",
        "#D0585E",
        "#CC4950",
        "#C73A41",
        "#BB2D34",
        "#9B262B",
        "#7C1E23",
        "#5C171A",
        "#3C0F11",
        "#1D0809"
      )
    col <- colorRampPalette(palet)(nclass)
  }
  if (missing(shift)) {
    shift <- diff(st_bbox(x)[c(1, 3)]) / 700
  }
  x <- x[order(x$min), ]
  plot(st_geometry(x),
    col = NA,
    border = NA,
    add = add
  )
  for (i in seq_len(nrow(x))) {
    p <- st_geometry(x[i, ])
    plot(
      p + c(-shift, shift),
      col = light,
      border = light,
      add = TRUE
    )
    plot(
      p + c(shift * 5 / 3, -shift * 5 / 3),
      col = dark,
      border = dark,
      add = TRUE
    )
    plot(p,
      col = col[i],
      border = "NA",
      add = TRUE
    )
  }
  leg(
    type = "choro",
    pos = legend.pos,
    val = c(x$min, max(x$max)),
    val_rnd = 0,
    pal = col,
    title = legend.title,
    size = 0.8, title_cex = .8, val_cex = .6,
  )
}
